File: native/chrome_openwith_mpv.py
#!/usr/bin/env python3
import json
import os
from pathlib import Path
import struct
import subprocess
import sys
from urllib.parse import urlparse


YOUTUBE_CONFIG_DIR = Path.home() / ".config" / "chrome-openwith-mpv" / "mpv-youtube"
YOUTUBE_CONFIG_FILE = YOUTUBE_CONFIG_DIR / "mpv.conf"
MPV_CONFIG_FILE = Path.home() / ".config" / "mpv" / "mpv.conf"


def is_youtube_url(url):
    if not isinstance(url, str) or not url.startswith(("http://", "https://")):
        return False

    host = urlparse(url).hostname or ""
    return host == "youtu.be" or host.endswith("youtube.com") or host.endswith("youtube-nocookie.com")


def ensure_youtube_profile():
    YOUTUBE_CONFIG_DIR.mkdir(parents=True, exist_ok=True)
    if YOUTUBE_CONFIG_FILE.exists():
        return

    if not MPV_CONFIG_FILE.exists():
        YOUTUBE_CONFIG_FILE.write_text("", encoding="utf-8")
        return

    lines = MPV_CONFIG_FILE.read_text(encoding="utf-8").splitlines()
    filtered = [line for line in lines if not line.lstrip().startswith("ytdl-raw-options=")]
    YOUTUBE_CONFIG_FILE.write_text("\n".join(filtered) + ("\n" if filtered else ""), encoding="utf-8")


def build_launch_args(url, headers=None, start_time=None):
    mpv = os.environ.get("MPV_BIN", "mpv")
    args = [mpv]

    if is_youtube_url(url):
        ensure_youtube_profile()
        args.append(f"--config-dir={YOUTUBE_CONFIG_DIR}")
    elif headers:
        header_fields = ", ".join(f"{name}: {value}" for name, value in headers.items())
        args.append(f"--http-header-fields={header_fields}")

    if start_time is not None:
        args.append(f"--start={start_time}")

    args.append(url)
    return args


def wait_for_mpv_window(pid, timeout_seconds=15, runner=None):
    if not os.environ.get("DISPLAY"):
        return True

    if runner is None:
        runner = subprocess.run

    command = [
        "xdotool",
        "search",
        "--sync",
        "--onlyvisible",
        "--pid",
        str(pid),
        "--limit",
        "1",
    ]

    try:
        result = runner(
            command,
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            timeout=timeout_seconds,
            check=False,
        )
        return result.returncode == 0
    except FileNotFoundError:
        return True
    except subprocess.TimeoutExpired:
        return False


def handle_message(message, launcher=None):
    url = message.get("url", "")
    headers = message.get("headers") or {}
    start_time = message.get("start_time")
    if not isinstance(url, str) or not url.startswith(("http://", "https://")):
        return {"ok": False, "error": "unsupported_url"}

    if launcher is None:
        launcher = launch_mpv
    if start_time is None:
        launcher(url, headers=headers)
    else:
        launcher(url, headers=headers, start_time=start_time)
    return {"ok": True}


def launch_mpv(url, headers=None, start_time=None, popen=None, window_waiter=None):
    return launch_mpv_with_process_wait(
        url,
        headers=headers,
        start_time=start_time,
        popen=popen,
        window_waiter=window_waiter,
    )


def launch_mpv_with_process_wait(url, headers=None, start_time=None, popen=None, window_waiter=None):
    args = build_launch_args(url, headers=headers, start_time=start_time)
    if popen is None:
        popen = subprocess.Popen
    if window_waiter is None:
        window_waiter = wait_for_mpv_window

    process = popen(
        args,
        stdin=subprocess.DEVNULL,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        start_new_session=True,
    )
    window_waiter(process.pid)
    return process


def read_message(stdin):
    raw_length = stdin.read(4)
    if not raw_length:
        return None
    if len(raw_length) != 4:
        raise ValueError("incomplete_length")
    length = struct.unpack("<I", raw_length)[0]
    payload = stdin.read(length)
    if len(payload) != length:
        raise ValueError("incomplete_payload")
    return json.loads(payload.decode("utf-8"))


def write_message(stdout, message):
    payload = json.dumps(message).encode("utf-8")
    stdout.write(struct.pack("<I", len(payload)))
    stdout.write(payload)
    stdout.flush()


def run(stdin=None, stdout=None, launcher=None):
    stdin = stdin or sys.stdin.buffer
    stdout = stdout or sys.stdout.buffer
    try:
        message = read_message(stdin)
        if message is None:
            return 0
        response = handle_message(message, launcher=launcher)
    except Exception as exc:
        response = {"ok": False, "error": str(exc)}
    write_message(stdout, response)
    return 0


if __name__ == "__main__":
    raise SystemExit(run())

File: native/chrome_openwith_mpv.js
#!/usr/bin/env node
const fs = require("node:fs");
const os = require("node:os");
const path = require("node:path");
const { spawn, spawnSync } = require("node:child_process");

const YOUTUBE_CONFIG_DIR = path.join(os.homedir(), ".config", "chrome-openwith-mpv", "mpv-youtube");
const YOUTUBE_CONFIG_FILE = path.join(YOUTUBE_CONFIG_DIR, "mpv.conf");
const MPV_CONFIG_FILE = path.join(os.homedir(), ".config", "mpv", "mpv.conf");

function isHttpUrl(url) {
  return typeof url === "string" && (url.startsWith("http://") || url.startsWith("https://"));
}

function isYoutubeUrl(url) {
  if (!isHttpUrl(url)) {
    return false;
  }

  const host = new URL(url).hostname || "";
  return host === "youtu.be" || host.endsWith("youtube.com") || host.endsWith("youtube-nocookie.com");
}

function ensureYoutubeProfile() {
  fs.mkdirSync(YOUTUBE_CONFIG_DIR, { recursive: true });

  if (fs.existsSync(YOUTUBE_CONFIG_FILE)) {
    return;
  }

  if (!fs.existsSync(MPV_CONFIG_FILE)) {
    fs.writeFileSync(YOUTUBE_CONFIG_FILE, "", "utf8");
    return;
  }

  const lines = fs.readFileSync(MPV_CONFIG_FILE, "utf8").split(/\r?\n/);
  const filtered = lines.filter((line) => !line.trimStart().startsWith("ytdl-raw-options="));
  fs.writeFileSync(YOUTUBE_CONFIG_FILE, `${filtered.join("\n")}${filtered.length ? "\n" : ""}`, "utf8");
}

function buildLaunchArgs(url, headers = null, start_time = null) {
  const mpv = process.env.MPV_BIN || "mpv";
  const args = [mpv];

  if (isYoutubeUrl(url)) {
    ensureYoutubeProfile();
    args.push(`--config-dir=${YOUTUBE_CONFIG_DIR}`);
  } else if (headers && Object.keys(headers).length > 0) {
    const headerFields = Object.entries(headers).map(([name, value]) => `${name}: ${value}`).join(", ");
    args.push(`--http-header-fields=${headerFields}`);
  }

  if (start_time !== null && start_time !== undefined) {
    args.push(`--start=${start_time}`);
  }

  args.push(url);
  return args;
}

function waitForMpvWindow(pid, timeoutSeconds = 15, runner = spawnSync) {
  if (!process.env.DISPLAY) {
    return true;
  }

  const result = runner(
    "xdotool",
    ["search", "--sync", "--onlyvisible", "--pid", String(pid), "--limit", "1"],
    { stdio: "ignore", timeout: timeoutSeconds * 1000 }
  );

  if (result?.error?.code === "ENOENT") {
    return true;
  }

  if (result?.error?.name === "TimeoutError") {
    return false;
  }

  return result?.status === 0;
}

function launchMpvWithProcessWait(url, headers = null, start_time = null, popen = spawn, window_waiter = waitForMpvWindow) {
  const args = buildLaunchArgs(url, headers, start_time);
  const child = popen(args[0], args.slice(1), {
    stdio: "ignore",
    detached: true,
  });

  window_waiter(child.pid, 15);
  return child;
}

function launchMpv(url, headers = null, start_time = null, popen = spawn, window_waiter = waitForMpvWindow) {
  return launchMpvWithProcessWait(url, headers, start_time, popen, window_waiter);
}

function handleMessage(message, launcher = launchMpv) {
  const url = message?.url || "";
  const headers = message?.headers || {};
  const start_time = message?.start_time;

  if (!isHttpUrl(url)) {
    return { ok: false, error: "unsupported_url" };
  }

  if (start_time === null || start_time === undefined) {
    launcher(url, headers);
  } else {
    launcher(url, headers, start_time);
  }

  return { ok: true };
}

function readMessage(source = 0) {
  const read = (length) => {
    if (typeof source === "number") {
      const buffer = Buffer.alloc(length);
      let offset = 0;
      while (offset < length) {
        const bytesRead = fs.readSync(source, buffer, offset, length - offset, null);
        if (bytesRead === 0) {
          return offset === 0 ? null : buffer.subarray(0, offset);
        }
        offset += bytesRead;
      }
      return buffer;
    }

    if (typeof source.read === "function") {
      return source.read(length);
    }

    throw new Error("unsupported_input_source");
  };

  const rawLength = read(4);
  if (!rawLength) {
    return null;
  }

  if (rawLength.length !== 4) {
    throw new Error("incomplete_length");
  }

  const length = rawLength.readUInt32LE(0);
  const payload = read(length);
  if (!payload || payload.length !== length) {
    throw new Error("incomplete_payload");
  }

  return JSON.parse(payload.toString("utf8"));
}

function writeMessage(target, message) {
  const payload = Buffer.from(JSON.stringify(message), "utf8");
  const length = Buffer.alloc(4);
  length.writeUInt32LE(payload.length, 0);

  if (typeof target === "number") {
    fs.writeSync(target, length);
    fs.writeSync(target, payload);
    return;
  }

  if (typeof target.write === "function") {
    target.write(length);
    target.write(payload);
    if (typeof target.flush === "function") {
      target.flush();
    }
    return;
  }

  throw new Error("unsupported_output_target");
}

function run(stdin = 0, stdout = 1, launcher = launchMpv) {
  try {
    const message = readMessage(stdin);
    if (message === null) {
      return 0;
    }

    const response = handleMessage(message, launcher);
    writeMessage(stdout, response);
  } catch (error) {
    writeMessage(stdout, { ok: false, error: String(error.message || error) });
  }

  return 0;
}

if (require.main === module) {
  process.exitCode = run();
}

module.exports = {
  buildLaunchArgs,
  ensureYoutubeProfile,
  handleMessage,
  isHttpUrl,
  isYoutubeUrl,
  launchMpv,
  launchMpvWithProcessWait,
  readMessage,
  run,
  waitForMpvWindow,
  writeMessage,
};
